### ISSUE TITLE
Possible new enhancements?

### ISSUE THREAD

**Issue by truefalsename** [typora/typora-issues#125]
Hey, I have adopted Typora as my main notes app for scientific purposes. I believe it really has the potential to replace stuff like LyX and even Latex when writing down calculations and notes. Based on my experience so far, I came up with a small list of possible enhancements. I do not know if they can all be implemented, anyway I will write them down below:
- It would be nice if it was possible to configure mathjax for instance to change the default math font
- It would be awesome to have collapsable sections where to hide stuff like proofs of theorem and details of calculations that you do not want to see all the time
- tab completion for latex commands (something like eq + tab for \begin{equation} ... cursor here ... \end{equation}) would be sweet
- would it be possible to change the behaviours of brackets in a way that if I type in a bracket while some text is underlined I get a pair of brackets surrounding this text (currently it just substitutes the with the bracket)

cheers!

**Comment by waahto** [comment]
+1 for collapsible sections and brackets[1]!

[1] I'd also change the behaviour of brackets and quotes so that when the user deletes the first quote/bracket, it doesn't remove the second.
**| = cursor**
```
and next some quotes "|"
[backspace]
and next some quotes |  [current behaviour]
and next some quotes |" [preferred behaviour]
```

**Comment by abnerlee** [comment]
> It would be nice if it was possible to configure mathjax for instance to change the default math font

This is possible by supporting some configuration file

> It would be awesome to have collapsable sections where to hide stuff like proofs of theorem and details of calculations that you do not want to see all the time

If my understanding is correct, a section is a header and its following content, am I correct?

> It would be awesome to have collapsable sections where to hide stuff like proofs of theorem and details of calculations that you do not want to see all the time.

Sorry, but this seems not high priority and seems lots of work is needed, so won't be available before 1.0

> would it be possible to change the behaviours of brackets in a way that if I type in a bracket while some text is underlined I get a pair of brackets surrounding this text (currently it just substitutes the with the bracket)

If some text is selected, and user press bracket, then the brackets would surround the text, if no text is selected, then bracket will be inserted in the cursor position. I think it is the auto pair behavior for most text editors including typora

**Comment by leewi9** [comment]
It's good to have such add-on functionalities, but as you can see, Typora
still has some potential bugs needs to be addressed.

  

From my humble opinion, it would be better to have such bugs fixed before we
take further steps.  
  
Sent from [Nylas N1](https://nylas.com/n1?ref=n1), the extensible, open source
mail client.  

On Apr 19 2016, at 2:58 pm, AbnerLee &lt;notifications@github.com&gt; wrote:  

> > It would be nice if it was possible to configure mathjax for instance to
change the default math font

>

> This is possible by supporting some configuration file

>

>> It would be awesome to have collapsable sections where to hide stuff like
proofs of theorem and details of calculations that you do not want to see all
the time

>

> If my understanding is correct, a section is a header and its following
content, am I correct?

>

>> It would be awesome to have collapsable sections where to hide stuff like
proofs of theorem and details of calculations that you do not want to see all
the time.

>

> Sorry, but this seems not high priority and seems lots of work is needed, so
won't be available before 1.0

>

>> would it be possible to change the behaviours of brackets in a way that if
I type in a bracket while some text is underlined I get a pair of brackets
surrounding this text (currently it just substitutes the with the bracket)

>

> If some text is selected, and user press bracket, then the brackets would
surround the text, if no text is selected, then bracket will be inserted in
the cursor position. I think it is the auto pair behavior for most text
editors including typora

>

> —  
You are receiving this because you are subscribed to this thread.  
Reply to this email directly or [view it on GitHub](https://github.com/typora
/typora-issues/issues/125#issuecomment-211761387)![](https://github.com/notifi
cations/beacon/AKVKoBtjmHNKjwrMTbBYZ1Sv6AVpO-Kkks5p5H0NgaJpZM4IJlsh.gif)

**Comment by abnerlee** [comment]
@waahto Actually I don't want typora's auto complete for pair quotes/brackets be different from other text editor, so that users can switch between different code/markdown editors or IDEs for different works smoothly. 

And I checked other text editors like Sublime, it behaves like current Typora, so I prefer to keep current behaviors unless you got some other concerns

**Comment by truefalsename** [comment]
> This is possible by supporting some configuration file

cool! 

>If my understanding is correct, a section is a header and its following content, am I correct?

yes, you could live the header out exactly as in mathemathica. I am not in rush and can wait until v1.0, meanwhile I will advertise typora amongst theoretical physicists as the best note editor ever ;)

> If some text is selected, and user press bracket, then the brackets would surround the text, if no text is selected, then bracket will be inserted in the cursor position. I think it is the auto pair behavior for most text editors including typora

You are right and I must have been drunk.

> change the default math font
Is there a way to do that at the current stage? In latex you would use a package like cmbright but in mathjax, as far as I understand, you have to modify the settings file which is hidden somewhere inside the Typora package.

**Comment by truefalsename** [comment]
One very las thing: I am toying with the Github CSS style modifying some fonts colors etc.  When I visualise the outline panel there is something like some light gradient that separates it from the main page. How do I get rid of that? I see that other styles have a simple line there but cannot track down the part of code in the CSS style file that sets this element

**Comment by waahto** [comment]
>@waahto Actually I don't want typora's auto complete for pair quotes/brackets be different from other text editor, so that users can switch between different code/markdown editors or IDEs for different works smoothly.

Yeah, better keep it consistent. I just turned the auto pairing off.

> From my humble opinion, it would be better to have such bugs fixed before we
take further steps.

I agree on fixing the bugs before adding new features. There's still a bunch of bugs that are pretty lame; for example #47 and #41 which are quite old even.

**Comment by sathomas** [comment]
Would definitely love to see a configuration option for MathJax

**Comment by masonlr** [comment]
A MathJax configuration file would be incredibly useful for scientific writing, i.e. the ability to include Macros (http://docs.mathjax.org/en/latest/tex.html#defining-tex-macros) and the MathJax extensions (http://docs.mathjax.org/en/latest/tex.html#tex-and-latex-extensions) that live in `Typora.app/Contents/Resources/TypeMark/lib/MathJax-2.6.1/extensions/TeX`

I can see Typora being incredibly useful for researchers who use markdown/pandoc/mathjax but want to make it seamless for collaborators to contribute who would otherwise only be comfortable with MS Word. And, to avoid Lyx as mentioned by @truefalsename.

It would be brilliant if constructs such as

```
$$
a + b = c
$$ {#eq:label}
```

didn't break the display of equations in Typora. (See the comparison in the image to atom  markdown-preview-plus). 

![](https://cloud.githubusercontent.com/assets/5828176/23236377/0fbcfa40-f9ad-11e6-92da-d7123946733e.png)

The absence of rendered cross-referencing (eg. via `pandoc --filter=pandoc-crossref, --filter=pandoc-citeproc, --bibliography=bibliography.bib`) is okay, but the breaking equations is preventing me from using the Typora app.

FYI, I came across Typora by reading this interesting scientific workflow idea (https://github.com/mschroen/Science.md).

**Comment by masonlr** [comment]
> A MathJax configuration file would be incredibly useful for scientific writing, i.e. the ability to include Macros (http://docs.mathjax.org/en/latest/tex.html#defining-tex-macros) and the MathJax extensions (http://docs.mathjax.org/en/latest/tex.html#tex-and-latex-extensions) that live in Typora.app/Contents/Resources/TypeMark/lib/MathJax-2.6.1/extensions/TeX

Ah sorry, I can see that this is already possible by editing `MathJax.Hub.Config` in `/Applications/Typora.app/Contents/Resources/TypeMark/index.html`. Brilliant.

**Comment by abnerlee** [comment]
I will separate the proposal here

- configure mathjax -> https://github.com/typora/typora-issues/issues/100
- collapsable sections -> https://github.com/typora/typora-issues/issues/499
- behaviours of brackets -> supported already
